### ISSUE TITLE
bottlerocket custom cluster-dns-ip doesn't work in ec2nodeclass userdata

### ISSUE THREAD

**Issue by nikoul** [aws/karpenter-provider-aws#5584]
### Description

**Observed Behavior**:

When using the following `EC2NodeClass`, karpenter should set `cluster-dns-ip` to `169.254.5.5` (that's a random ip address I chose to use with [nodelocaldns](https://kubernetes.io/docs/tasks/administer-cluster/nodelocaldns/)):
```
apiVersion: karpenter.k8s.aws/v1beta1
kind: EC2NodeClass
metadata:
  name: karpenter-cpu
spec:
  amiFamily: Bottlerocket
  amiSelectorTerms:
  - name: bottlerocket-aws-k8s-1.27-x86_64-*
  ....
  userData: |
    [settings.kubernetes]
    cluster-name = "infra-dev-1"
    cluster-dns-ip = "169.254.5.5"
```

However, after karpenter created a Bottlerocket node based on that `EC2NodeClass`, here is the resulting UserData (as seen in the EC2 console) : 

```
[settings]
[settings.kubernetes]

cluster-name = 'infra-dev-1'
cluster-dns-ip = '172.20.0.10'
```

**Expected Behavior**:
karpenter should set the `cluster-dns-ip` to the specified value. I don't think there's a particular issue with my setup, as other values like `cluster-name` are set correctly based on what was specified in the `EC2NodeClass`.

**Versions**:
- Chart Version: 0.32.3
- Kubernetes Version (`kubectl version`): 1.27 (EKS)

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Comment by jonathan-innis** [comment]
I agree that it's a bit odd that Karpenter will override a field that you specified in your custom UserData with a discovered value. Right now, what's happening is that Karpenter is trying to discover ClusterDNS for the cluster that Karpenter is running on and use that value when launching new nodes (see https://github.com/aws/karpenter-provider-aws/blob/main/pkg/providers/amifamily/resolver.go#L192). 

Realistically, I do agree that we should probably respect values in the following order of precedence:
1. Values specified in your custom userData
2. Values specified in your kubelet config
3. Default values discovered by Karpenter.

As a workaround for now, I'd recommend that you set the ClusterDNS through the KubeletConfig in the `spec.template.spec.kubelet.clusterDNS` field.

**Comment by nikoul** [comment]
> I agree that it's a bit odd that Karpenter will override a field that you specified in your custom UserData with a discovered value. Right now, what's happening is that Karpenter is trying to discover ClusterDNS for the cluster that Karpenter is running on and use that value when launching new nodes (see https://github.com/aws/karpenter-provider-aws/blob/main/pkg/providers/amifamily/resolver.go#L192).
> 
> Realistically, I do agree that we should probably respect values in the following order of precedence:
> 
> 1. Values specified in your custom userData
> 2. Values specified in your kubelet config
> 3. Default values discovered by Karpenter.
> 
> As a workaround for now, I'd recommend that you set the ClusterDNS through the KubeletConfig in the `spec.template.spec.kubelet.clusterDNS` field.

Thank you for the explanation and the workaround. It works fine by setting the ClusterDNS in the NodePool's kubelet config 👍🏻

**Comment by lteixeira-dock** [comment]
@jonathan-innis I have a similar error, but In my case the parameter is in EKS Cluster layer. The documentation say that Karpenter override this field, but In this case is overriding wrong.

I opened a issue for it: https://github.com/aws/karpenter-provider-aws/issues/5702

Can you help me, please?

**Comment by jonathan-innis** [comment]
@nikoul Do you know why the Karpenter discovered value was wrong in your case? It sounds like there was an issue with the service IP that was discovered? Specifically, it sounds like you have the `kube-dns` service that was running on `172.20.0.10` in the cluster and that's what Karpenter passed through, though somehow that's inaccurate with the actual state of your cluster. Do you know why this value isn't accurate for you?

**Comment by z0rc** [comment]
@jonathan-innis given that `169.254.5.5` is from link-local unicast cidr, it's most likely @nikoul wants to set up node-local-dns. At least I got similar problem with inability to override cluster-dns-ip via userData, which lead me here.

node-local-dns runs as deamonset, uses `kube-dns` as upstream and provides dns cache for node. In this configuration kube-dns is needed and present in cluster, but it isn't used by pods directly.

**Comment by dhanutalari** [comment]
Hi 
can someone help me resolve the issue that my nodeclaim not getting initize duo to ec2nodeclass userdata below
userData: |
   [settings.kernel.sysctl]
   net.ipv4.conf.all.send_redirects = 0

Here i need to change the sysctl variable values using userdata .

**Comment by diranged** [comment]
In general - I was surprised to find that Karpenter _overwrites_ settings we pass in via UserData. I feel like that anything the user supplies should become the ultimate setting, without exception. Is there room for a discussion on this behavior?

**Comment by mzupan** [comment]
This is with karpenter 1.8.2 and bottlerocket 1.46.0

```
spec:
    kubelet:
      clusterDNS:
        - "169.254.20.10"
        - "172.20.0.10"
```

The /etc/resolv,conf for pods on the node is only  `169.254.20.10`

Its better then not using node-local but still not ideal
